common-rest/src/main/java/org/uniprot/api/rest/validation/ValidPostByIdsRequest.java

```diff
@@ -0,0 +1,147 @@
+package org.uniprot.api.rest.validation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.regex.Pattern;
+
+import javax.validation.Constraint;
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+import javax.validation.Payload;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.commons.beanutils.BeanUtils;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.web.context.support.SpringBeanAutowiringSupport;
+import org.uniprot.core.util.Utils;
+import org.uniprot.store.config.UniProtDataType;
+import org.uniprot.store.config.returnfield.config.ReturnFieldConfig;
+import org.uniprot.store.config.returnfield.factory.ReturnFieldConfigFactory;
+
+/**
+ * @author sahmad
+ * @created 26/07/2021
+ */
+@Constraint(validatedBy = ValidPostByIdsRequest.PostByIdsRequestValidator.class)
+@Target({ElementType.TYPE, ElementType.ANNOTATION_TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface ValidPostByIdsRequest {
+    String message() default "invalid ids download request";
+
+    String accessions() default "accessions";
+
+    String download() default "download";
+
+    String fields() default "fields";
+
+    UniProtDataType uniProtDataType();
+
+    Class<?>[] groups() default {};
+
+    Class<? extends Payload>[] payload() default {};
+
+    @Slf4j
+    class PostByIdsRequestValidator extends CommonIdsRequestValidator
+            implements ConstraintValidator<ValidPostByIdsRequest, Object> {
+
+        @Value("${ids.max.download.length}")
+        private String maxDownloadLength;
+
+        private UniProtDataType uniProtDataType;
+        private String accessions;
+        private String download;
```

Comment:
I think we could remove this download validation

Reply:
Hi Leo,
Initially, i was using the same validator ValidUniqueIdList for the Post request but it was becoming messier. So i wrote a separate validator to keep the Post request validation clean. Please notice that we are validating the whole Request object at once. This makes sense for POST Json body request. For GET request where we send query param, it makes more sense to validate each field individually.
If you see there are few other differences in ValidPostByIdsRequest and ValidUniqueIdList :
- ValidUniqueIdList validates the length by ids.max.length but ValidPostByIdsRequest validates ids.max.download.length
- We are not validating empty or null accessions in ValidUniqueIdList but we are validating these conditions in ValidPostByIdsRequest
- The error message we return from ValidPostByIdsRequest is different than ValidUniqueIdList.
Yes, i will remove the validation on download field.
Let me know what do you think?